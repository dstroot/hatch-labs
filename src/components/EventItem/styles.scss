.react-add-to-calendar__dropdown ul {
  padding-left: 0;
}
