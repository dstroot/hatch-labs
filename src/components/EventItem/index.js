import React from "react";
import "./styles.scss";

// https://jasonsalzman.github.io/react-add-to-calendar/
import AddToCalendar from "react-add-to-calendar";
import "react-add-to-calendar/dist/react-add-to-calendar.css";

// we will render one "item" for each calendar record we recieve from
// our API
const EventItem = ({
  name,
  description,
  date,
  start,
  end,
  address1,
  address2,
  city,
  state,
  zip,
  latitude,
  longitude,
  imageURL
}) => {
  let event = {
    title: name,
    description: description,
    location: address1,
    startTime: start,
    endTime: end
  };

  let items = [
    { outlook: "Microsoft" },
    { apple: "Apple" },
    { google: "Google" }
  ];

  let icon = { "calendar-plus-o": "left" };

  return (
    <li className="media mt-4">
      <img
        className="mr-3 d-none d-sm-block"
        src={imageURL[0].url}
        width="200px"
        height="225px"
        alt="event"
      />
      <div className="media-body align-self-center">
        <div className="row">
          <div className="col-md-8">
            <h4 className="mt-0 mb-1">{name}</h4>
            <p className="text-muted small mb-2">
              {new Date(date).toDateString()}{" "}
              {new Date(start).toLocaleTimeString("en-US", {
                hour: "numeric",
                hour12: true,
                minute: "numeric"
              })}{" "}
              {"- "}
              {new Date(end).toLocaleTimeString("en-US", {
                hour: "numeric",
                hour12: true,
                minute: "numeric"
              })}
            </p>
          </div>
          <div className="col-md-4">
            <AddToCalendar
              event={event}
              listItems={items}
              buttonTemplate={icon}
              buttonClassClosed="btn btn-primary dropdown-toggle"
              buttonClassOpen="btn btn-primary dropdown-toggle"
              buttonLabel=" > My Calendar"
              buttonWrapperClass="dropdown mb-3"
            />
          </div>
        </div>
        <p className="">{description}</p>
        <address className="small text-muted">
          {address1}, {address2}
          <br />
          {city}, {state} {zip}
        </address>
      </div>
    </li>
  );
};

export default EventItem;
