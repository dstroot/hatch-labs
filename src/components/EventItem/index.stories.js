import React from "react";
import { storiesOf } from "@storybook/react";
import EventItem from "../EventItem";

storiesOf("EventItem", module).add("show EventItem", () => <EventItem />);
