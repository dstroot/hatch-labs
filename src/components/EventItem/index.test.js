import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import EventItem from "../EventItem";

const data = {
  name: "Name",
  description: "Description",
  date: "2019-01-31",
  start: "2019-01-31T20:00:00.000Z",
  end: "2019-01-31T21:30:00.000Z",
  address1: "Address One",
  address2: "Address Two",
  city: "City",
  state: "State",
  zip: "99999",
  latitude: "",
  longitude: "",
  imageURL: [
    {
      filename: "hatch.png",
      id: "attSuiujHyZSyOqOP",
      size: 3212,
      thumbnails: [],
      type: "image/png",
      url:
        "https://dl.airtable.com/.attachments/876151ddacba830c5eab7acb2daffafb/ee66c7e9/hatch.png"
    }
  ]
};

describe("EventItem", () => {
  // mount the component
  let mountedComponent;
  const getMountedComponent = () => {
    if (!mountedComponent) {
      mountedComponent = mount(<EventItem {...data} />);
    }
    return mountedComponent;
  };

  const RealDate = Date;

  beforeEach(() => {
    mountedComponent = undefined;

    // mock the date/time
    const currentDate = new Date("2019-01-31T20:00:00.000Z");
    global.Date = jest.fn(() => new RealDate(currentDate.toISOString()));
    Object.assign(Date, RealDate);
  });

  afterEach(() => {
    global.Date = RealDate;
  });

  it("it should render", () => {
    // mockDate("2017-11-25T12:34:56z");
    const component = renderer.create(<EventItem {...data} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("it should say `Name`", () => {
    const h4 = getMountedComponent()
      .find("h4")
      .first();

    expect(h4.text()).toContain("Name");
  });
});

// Testing information
// ---------------------------------------------------
// Jest - Expect: https://jestjs.io/docs/en/expect
// https://jestjs.io/docs/en/expect.html#content
// Enzyme cheatsheet - https://devhints.io/enzyme

// https://medium.com/@stipsan/testing-with-jest-15-awesome-tips-and-tricks-42150ec4c262
// https://www.youtube.com/watch?time_continue=459&v=XUlGzJLZe2Q
// https://medium.freecodecamp.org/the-right-way-to-test-react-components-548a4736ab22
// https://blog.pragmatists.com/click-your-element-how-to-make-enzyme-tests-cleaner-3e7ad21c2be0
// https://blog.bitsrc.io/how-to-test-react-components-with-jest-and-enzyme-in-depth-145fcd06b90
// https://medium.com/front-end-hacking/tested-react-lets-build-a-data-table-a76aa100d23f

// Component Contracts
// --------------------------------------------------
// In order to test a component you must first understand what its
// Contract is. Understanding a component’s contract is the most important
// part of testing a React component. A contract defines the expected behavior
// of your component and what assumptions are reasonable to have about its
// usage. Without a clear contract, your component may be hard to understand.
// Writing tests is a great way to formally define your component’s contract.

// Every React component has at least one thing that contributes to the
// definition of its contract:

// - What it renders (which may be nothing)
//
// Additionally, most component contracts are affected by these things as well:
//
// - The props the component receives
// - The state the component holds (if any)
// - What the component does when the user interacts with it
//   (via clicking, dragging, keyboard input, etc)
