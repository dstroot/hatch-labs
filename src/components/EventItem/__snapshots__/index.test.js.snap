// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventItem it should render 1`] = `
<li
  className="media mt-4"
>
  <img
    alt="event"
    className="mr-3 d-none d-sm-block"
    src="https://dl.airtable.com/.attachments/876151ddacba830c5eab7acb2daffafb/ee66c7e9/hatch.png"
  />
  <div
    className="media-body align-self-center"
  >
    <div
      className="row"
    >
      <div
        className="col-md-8"
      >
        <h4
          className="mt-0 mb-1"
        >
          Name
        </h4>
        <p
          className="text-muted small mb-2"
        >
          Thu Jan 31 2019
           
          12:00 PM
           
          - 
          12:00 PM
        </p>
      </div>
      <div
        className="col-md-4"
      >
        <div
          className="react-add-to-calendar"
        >
          <div
            className="dropdown mb-3"
          >
            <a
              className="btn btn-primary dropdown-toggle"
              onClick={[Function]}
            >
              <span>
                <i
                  className="react-add-to-calendar__icon--left fa fa-calendar-plus-o"
                />
                  &gt; My Calendar
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
    <p
      className=""
    >
      Description
    </p>
    <address
      className="small text-muted"
    >
      Address One
      , 
      Address Two
      <br />
      City
      , 
      State
       
      99999
    </address>
  </div>
</li>
`;
